[9.0][MIG] purchase product variant configurator

zakiuu opened the pull request:
Migration of product_variant_configurator and purchase_product_variant_configurator modules

oihane committed:
Renamed product_variants modules

oihane committed:
New name and security files fixed

oihane committed:
New name and security files fixed

oihane committed:
<purchase_product_variants> Error with onchange solved

oihane committed:
<purchase_product_variants> Lil' change

oihane committed:
<purchase_product_variants> PEP8 error fixed

oihane committed:
<product_variants_no_automatic_creation> _product_find

oihane committed:
Fix pep8

oihane committed:
possible values added to scheduled lines for attributes

oihane committed:
Attributes in mo product lines

oihane committed:
Operation assignation fixing

oihane committed:
Fixed errors and changes

oihane committed:
translations

oihane committed:
get inherit attribute value, and corresponding product if exists

oihane committed:
<product_variants_no_automatic_creation> prices according to attributes

oihane committed:
Fixing pricelist aplication

oihane committed:
PEP8

oihane committed:
import missing

oihane committed:
<product_variants_no_automatic_creation> view don't allow editing

oihane committed:
Update product_view.xml

oihane committed:
Change field attribute copiable to copy

oihane committed:
Not refreshing view after duplicating line fixed

pedrobaeza committed:
product_variants_no_automatic_creation: Enhance labels and defaults

pedrobaeza committed:
product_variants_no_automatic_creation: Labels

oihane committed:
New PEP8 error fixing

oihane committed:
variants creation fixed

oihane committed:
Avoid launching onchange with default value

pedrobaeza committed:
purchase_product_variants: Purchase order line product attributes put in another place

oihane committed:
Sale_variants description changed, and uom in purchase when template selected

oihane committed:
Sale_variants description changed, and uom in purchase when template selected

oihane committed:
Avoid template deleting

pedrobaeza committed:
Expand AUTHORS

pedrobaeza committed:
Expand AUTHORS

pedrobaeza committed:
purchase_product_variants: Limit attribute values

pedrobaeza committed:
*_product_variants: Closes #752. Button duplicate line not visible when order is confirmed

pedrobaeza committed:
*_product_variants: Copy button visibility

pedrobaeza committed:
product_variants_no_automatic_creation: Preserve order on attributes + improve code

pedrobaeza committed:
purchase_product_variants: Pass on onchange_product_id the one2many dictionary as expected in create method for bids

pedrobaeza committed:
purchase_product_variants: Create product variants before other checks

pedrobaeza committed:
purchase_product_variants: Bring data when changing template

pedrobaeza committed:
product_variants_no_automatic_creation: Make product template search usable without variants. Closes #804

pedrobaeza committed:
product_variants_no_automatic_creation: Adapt _product_find for looking for non full-set of values

pedrobaeza committed:
purchase_product_variants: Change purchase order line description while filling attribute values

pedrobaeza committed:
purchase_product_variants: Translation template + es

pedrobaeza committed:
purchase_product_variants: Lookup of existing product when confirming fixed

ddiazpinto committed:
purchase_product_variants: Checks the number of attributes selected in `purchase.order.line` and the corresponding `product.product` to ensure to be equal and avoid `expected singleton` error.

sysadminmatmoz committed:
Missing templates and translations added

sysadminmatmoz committed:
Slovene translations of odoomrp modules

pedrobaeza committed:
product_variants_no_automatic_creation: Include constraint for product/attribute uniqueness

oihane committed:
<purchase_product_variants> Update product_template with post-init-hook

oihane committed:
<purchase_product_variants> Required changes, updated README

pedrobaeza committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
purchase_product_variants_types

Product variants types (Sale)
=============================

This module extends product variants on sales. It adds the possibility of
defining a custom value when the attribute is of range type.

pedrobaeza committed:
product_variants_no_automatic_creation: Find correctly a product for its attributes

Adding to the domain all values that exists, not only the ones that have more than one
possible value. That was an old constraint because attributes with only one value were
not populated, but that was fixed a long time ago and now it makes that the match doesn't
happen because the comparison of the attributes length fail.

pedrobaeza committed:
*_product_variants: Respect attribute order

pedrobaeza committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
purchase_product_variants: Handle onchange_product_template properly

pedrobaeza committed:
*_product_variants: remove previous attribute values on product configurator

pedrobaeza committed:
product_variants_no_automatic_creation: Add configurator in product variant form

pedrobaeza committed:
*_product_variants: don't error on lines without product nor template

pedrobaeza committed:
product_variants_no_automatic_creation: Security for product.product.attribute

pedrobaeza committed:
product_variants_no_automatic_creation: Don't error on missing attribute values

pedrobaeza committed:
product_variants_no_automatic_creation: Avoid to disturb tests results

pedrobaeza committed:
product_variants_no_automatic_creation: Fix security file

jcarlosmontoya committed:
Update term es.po

lk-eska committed:
adds sale and purchase description to name fields

lk-eska committed:
remove unnecessary empty line

pedrobaeza committed:
purchase_product_variants: Propagate product template on MTO

pedrobaeza committed:
purchase_product_variants: Missing inherit

rdeheele committed:
add purchase_ok domain on template field

pedrobaeza committed:
product_variant_*: Abstract product configurator + application on stock.move

pedrobaeza committed:
*_product_variants: Some adaptations

pedrobaeza committed:
*_product_variants: Field names

pedrobaeza committed:
*_product_variants: Field name in views

pedrobaeza committed:
purchase_product_variants: Undo name change

levkar committed:
Add attrs and correct description on po creation (#1202)

[FIX] purchase_product_variants: Add attrs and correct description on po creation

oihane committed:
purchase_product_variants: update module with changes in base module

pedrobaeza committed:
*_product_variants_*: Fix migration scripts

pedrobaeza committed:
purchase_product_variants: Create correctly MTO orders

pedrobaeza committed:
product_variants_*: Handle translations in product description/attributes

pedrobaeza committed:
*_product_variants: Handle onchange product template

hugosantosred committed:
Fix _check_configuration_validity constraint

oihane committed:
product_variants_no_automatic_creation:

* Fix expected singleton error when updating attributes (#1237)
* Tests

pedrobaeza committed:
product_variants_no_automatic_creation

* Link quick-created attribute value to the template
* Don't require parent product_tmpl_id field in view

zakmak88 committed:
product_variant_configurator

This module has been ported from the odoomrp repo.
the module name in the odoo mrp repo was product_variants_no_automatic_creation
the following improvement where done while porting this module:

* Improve the readme file with a technical description on how the module should
  be used with other modules.
* Add attribute col=4 to group parent for a better desplay of product category
  for view.
* Remove unnecessary @api.multi before compute or onchange methods.
* Move product.configurator.attribute model to the new
  product_configurator_attribute.py file.
* Add unit test for product.configurator.attribute model that includes test for
  _compute_possible_value_ids and _compute_price_extra methods.
* Rename XML files in views and security folder as per OCA guidelines.
* Improve test coverage for product_configurator.
* Add unit test for purchase.order and purchase.order.line models.
* Separate the purchase.order and purchase.order.line models code into two files.
* Rename XML files as per OCA guidelines.
* product_variant_configurator: set product_tmpl_id in
  onchange_product_id_product_configurator_old_api and
  onchange_product_id_product_configurator methods for product.configurator model.
* Add unit test for onchange_product_tmpl_id and onchange_product_attribute_ids.

zakmak88 committed:
product_variant_configurator

This module has been ported from the odoomrp repo.
the module name in the odoo mrp repo was product_variants_no_automatic_creation
the following improvement where done while porting this module:

* Improve the readme file with a technical description on how the module should
  be used with other modules.
* Add attribute col=4 to group parent for a better desplay of product category
  for view.
* Remove unnecessary @api.multi before compute or onchange methods.
* Move product.configurator.attribute model to the new
  product_configurator_attribute.py file.
* Add unit test for product.configurator.attribute model that includes test for
  _compute_possible_value_ids and _compute_price_extra methods.
* Rename XML files in views and security folder as per OCA guidelines.
* Improve test coverage for product_configurator.
* Add unit test for purchase.order and purchase.order.line models.
* Separate the purchase.order and purchase.order.line models code into two files.
* Rename XML files as per OCA guidelines.
* product_variant_configurator: set product_tmpl_id in
  onchange_product_id_product_configurator_old_api and
  onchange_product_id_product_configurator methods for product.configurator model.
* Add unit test for onchange_product_tmpl_id and onchange_product_attribute_ids.

zakmak88 committed:
purchase_product_variants This module has been ported from the odoomrp repo. the following improvements were done while porting this module:

* Adapt files as per OCA guidelines.
* Add unit test

zakmak88 committed:
Rename purchase_product_variants > purchase_product_variant_configurator

zakmak88 committed:
Rename purchase_product_variants > purchase_product_variant_configurator

zakmak88 committed:
purchase_product_variants: Add test coverage to wkf_confirm_order function

zakmak88 committed:
purchase_product_variants: Add case where product_tmpl_id not yet set

zakmak88 committed:
product_variant configurator: Fix _price_rule_get_multi

Fix _price_rule_get_multi method inspired from v10 implementation
so the migration will be easier to v9 and v10

zakmak88 committed:
product_variant configurator: Unit test for _price_rule_get_multi

zakmak88 committed:
product_variant_configurator: ad sequence to pricelist test

zakmak88 committed:
product_variant_configurator: Remove duplicated xml files

zakmak88 committed:
Migrate to 9.0

chrisdec commented:
Functional tests : 
- When I create a new product, I have this error message : "Changing this parameter may cause automatic variants creation' --> This warning message must only appear if I change the variant creation field.
- Maybe it could be nice to add a menu "product variants" in purchase. I installed the sale module to have access to variants. 
-  I have a error message when I add a product without attribute to a purchase order. ( to reproduce : create a new product without attribute, create a supplier info, save and then, add this product to a purchase order). The error message is the following : 

### Odoo Server Error

Traceback (most recent call last):
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3270933-44-d50d01/openerp/http.py", line 648, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3270933-44-d50d01/openerp/http.py", line 685, in dispatch
    result = self._call_function(*_self.params)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3270933-44-d50d01/openerp/http.py", line 321, in _call_function
    return checked_call(self.db, *args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3270933-44-d50d01/openerp/service/model.py", line 118, in wrapper
    return f(dbname, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3270933-44-d50d01/openerp/http.py", line 314, in checked_call
    result = self.endpoint(_a, *_kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3270933-44-d50d01/openerp/http.py", line 964, in __call__
    return self.method(_args, *_kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3270933-44-d50d01/openerp/http.py", line 514, in response_wrap
    response = f(_args, *_kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3270933-44-d50d01/openerp/addons/web/controllers/main.py", line 889, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3270933-44-d50d01/openerp/addons/web/controllers/main.py", line 881, in _call_kw
    return getattr(request.registry.get(model), method)(request.cr, request.uid, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3270933-44-d50d01/openerp/api.py", line 250, in wrapper
    return old_api(self, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3270933-44-d50d01/openerp/api.py", line 381, in old_api
    result = method(recs, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3270933-44-d50d01/openerp/models.py", line 6070, in onchange
    record._onchange_eval(name, field_onchange[name], result)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3270933-44-d50d01/openerp/models.py", line 5927, in _onchange_eval
    method_res = method(self)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3270933-44-d50d01/openerp/addons/purchase_product_variant_configurator/models/purchase_order_line.py", line 38, in onchange_product_tmpl_id
    res = super(PurchaseOrderLine, self).onchange_product_tmpl_id()
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3270933-44-d50d01/openerp/addons/product_variant_configurator/models/product_configurator.py", line 49, in onchange_product_tmpl_id
    self.product_tmpl_id.product_variant_ids[0].id
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3270933-44-d50d01/openerp/models.py", line 5760, in __getitem__
    return self._browse(self.env, (self._ids[key],))
IndexError: tuple index out of range 
- Price on the supplier info is not taken into account when you work with attributes. (I tested the template supplier info and the variant supplier info, my price is always "0")
- The last attribute you encoded on the purchase order line is not taken into account. For example, I put a "veste" with color white and size "S" but only color white is registered. My created product variant is a "veste withe" and not a "veste white size s". I think it's a cursor problem. 

So, I stop here and I will restart tests on Wednesday :) Thanks @zakiuu !

zakmak88 committed:
[FIX]product_variant_configurator: create a default variant when no attribute is selected

zakmak88 committed:
[FIX]purchase_product_variant: onchange_product_tmpl_id
make onchange_product_tmpl_id behave exactly like onchange_product_id in the base module

zakmak88 committed:
[FIX]product_variant_configurator: onchange_product_tmpl_id in product.configurator
use smart commands to add product_attribute_ids

zakmak88 committed:
[FIX]purchase_product_variant_configurator: display variant inside a tab and make template required

zakmak88 committed:
[MIG]update module version purchase_product_variant_configurator and product_variant_configurator

zakmak88 committed:
[FIX]*product_variant_configurator:flake and duplicate xml

zakmak88 committed:
[FIX]purchase_product_variant_configurator: fix manifest

zakmak88 committed:
[MIG]product_variant_configurator: make module installable

zakmak88 committed:
[MIG]purchase_product_variant_configurator: make module installable

zakmak88 committed:
[FIX]product_variant_configurator: fix unit test for variant creation

zakmak88 committed:
[FIX]purchase_product_variant_configurator: bug fix in onchange_product_id and onchange_product_tmpl_id

sbidoul commented:
@zakiuu there are a few lint warnings in travis.
https://travis-ci.org/OCA/product-variant/jobs/171589902#L295

sbidoul commented on product_variant_configurator/models/product_category.py:
```diff
@@ -0,0 +1,36 @@
+                'title': _('Change warning!'),
+                'message': _('Changing this parameter may cause'
+                             ' automatic variants creation')
+            }}
```
This warning is always raised when opening the product creation form. We need to be smarter when deciding to raise the warning or not.

zakmak88 committed:
[FIX]*_product_variant_configurator: bug fix in onchange_product_id and onchange_product_id_product_configurator

zakmak88 committed:
[FIX]product_variant_configurator: Remove possible_value_ids and add an onchange method

zakmak88 committed:
[FIX]product_variant_configurator: Remove the call to _compute_possible_value_ids

zakmak88 committed:
[FIX]*_product_variant_configurator: remove possible_value_ids from attribute view

zakmak88 committed:
[FIX]*_product_variant_configurator: remove possible_value_ids from unit test and flake8 fix

zakmak88 committed:
[FIX]product_variant_configurator:Missing new line in product tmpl view

zakmak88 committed:
[FIX]product_variant_configurator:Fix flake 8 in pricelist.py

sbidoul commented on purchase_product_variant_configurator/models/purchase_order_line.py:
```diff
@@ -0,0 +1,70 @@
+        result = super(PurchaseOrderLine, self).onchange_product_tmpl_id()
+
+        self.date_planned = datetime.today().strftime(
+            DEFAULT_SERVER_DATETIME_FORMAT)
```
Use fields.Datetime.now(). 

That said, is it normal to use now() for date_planned?

zakiuu replied:
This is taken from the base

zakmak88 committed:
[REF] Refactore the use of the owner_id field

zakmak88 committed:
[REF]Refactore the use of the owner_id (fix unit tests)

zakmak88 committed:
[REF] Revert compute possible values

sbidoul committed:
[FIX] avoid systematic warning when creating new product templates

sbidoul commented on product_variant_configurator/models/product_configurator.py:
```diff
@@ -0,0 +1,254 @@
+                product.product_tmpl_id, product, product.attribute_value_ids)
+
+    @api.multi
+    def onchange_product_id_product_configurator_old_api(
```
I guess this method can be removed in v9?

pedrobaeza replied:
As this is a "framework", I would let it until v10 where old API is removed, although it has not been used in sale nor purchase modules.

sbidoul commented on product_variant_configurator/models/product_configurator.py:
```diff
@@ -0,0 +1,254 @@
+            self.product_id = \
+                self.product_tmpl_id.product_variant_ids[0].id
+        else:
+            if not self.env.context.get('not_reset_product'):
```
I could not find where not_reset_product is set. @zakiuu @pedrobaeza do you know what it's for?

pedrobaeza replied:
It's used here for avoiding a problem in the stock move: https://github.com/odoomrp/odoomrp-wip/blob/0bdffaa4be5e5607421b17cb1ad2df12e483f336/stock_product_variants/models/stock_move.py#L50

zakiuu commented:
Same haven't seen it in the code but maybe used in other modules on odoomrp
repo

On Sun, Oct 30, 2016, 12:47 Stéphane Bidoul (ACSONE) <
notifications@github.com> wrote:

> ## _@sbidoul_ commented on this pull request.
> 
> In product_variant_configurator/models/product_configurator.py
> https://github.com/OCA/product-variant/pull/44#pullrequestreview-6362491
> :
> 
> > -            record.price_extra = sum(
> > -                record.mapped('product_attribute_ids.price_extra'))
> >   +
> > -    @api.onchange('product_tmpl_id')
> > -    def onchange_product_tmpl_id(self):
> > -        if not self.product_tmpl_id:
> > -            return {}
> >   +
> > -        # First, empty current list
> > -        self.product_attribute_ids = [
> > -            (2, x.id) for x in self.product_attribute_ids]
> > -        if not self.product_tmpl_id.attribute_line_ids:
> > -            self.product_id = \
> > -                self.product_tmpl_id.product_variant_ids[0].id
> > -        else:
> > -            if not self.env.context.get('not_reset_product'):
> 
> I could not find where not_reset_product is set. @zakiuu
> https://github.com/zakiuu @pedrobaeza https://github.com/pedrobaeza
> do you know what it's for?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/OCA/product-variant/pull/44#pullrequestreview-6362491,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AANkfWlCygoFLQelFQ0rUu8XPH8ZWbr_ks5q5IPDgaJpZM4KjP_D
> .

sbidoul commented on product_variant_configurator/models/product_configurator.py:
```diff
@@ -0,0 +1,254 @@
+        return result
+
+    @api.model
+    def check_configuration_validity(self, vals):
```
@zakiuu this one is unused? Was it used in the original code?

sbidoul replied:
@pedrobaeza do you remember where this check_configuration_validity is supposed to be used?

pedrobaeza replied:
No, you can remove it. It's a trace of old code.

sbidoul commented on product_variant_configurator/models/product_configurator_attribute.py:
```diff
@@ -0,0 +1,57 @@
+
+    owner_id = fields.Integer(string="Owner",
+                              required=True,
+                              ondelete="cascade")
```
Does it make sense to have ondelete="cascade" on an Integer field?

sbidoul replied:
@zakiuu I understand why it is an Integer. What I don't understand is the ondelete=cascade. It has no effect, IMO.

zakiuu commented:
It's a reference to an abstract model, I didn't like it me too so I used a
many2many to not have to use it but Pedro showed me some examples where
that worked. So I've reverted it

On Sun, Oct 30, 2016, 13:54 Stéphane Bidoul (ACSONE) <
notifications@github.com> wrote:

> ## _@sbidoul_ commented on this pull request.
> 
> In product_variant_configurator/models/product_configurator_attribute.py
> https://github.com/OCA/product-variant/pull/44#pullrequestreview-6363292
> :
> 
> > +# © 2016 Pedro M. Baeza pedro.baeza@tecnativa.com
> > +# © 2016 ACSONE SA/NV
> > +# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3
> > +
> > +from openerp import api, fields, models
> > +from openerp.addons import decimal_precision as dp
> > +
> > +
> > +class ProductConfiguratorAttribute(models.Model):
> > -    _name = 'product.configurator.attribute'
> >   +
> > -    owner_model = fields.Char(required=True)
> >   +
> > -    owner_id = fields.Integer(string="Owner",
> > -                              required=True,
> > -                              ondelete="cascade")
> 
> Does it make sense to have ondelete="cascade" on an Integer field?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/OCA/product-variant/pull/44#pullrequestreview-6363292,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AANkffQHa9oXWjFh5pZHazzl4vD7c2apks5q5JN-gaJpZM4KjP_D
> .

sbidoul committed:
[FIX] fix warning when setting last attribute

What helps is setting the many2many_tags which seems to have an
influence even though the field is invisible. Without this you sometimes
get a warning asking if you want to discard the last change.
Setting the computed field readonly is also good practice.

sbidoul committed:
[IMP] new api style in onchange for updating attributes list

sbidoul commented:
@zakiuu @pedrobaeza I finally found a workaround (https://github.com/OCA/product-variant/pull/44/commits/748b3a6958e655395f0366e204b27cab0223f2a1) for the error with possible_value_ids so we can keep the computed field. Pfeew... not easy to find, and it's an Odoo web client bug IMHO.

Anyway, progress there is.

@zakiuu can you look after the failing test and the remaining comments (I'll add a few more questions)?

sbidoul committed:
[IMP] better to avoid overrides of onchange method (#2)

sbidoul commented on purchase_product_variant_configurator/__openerp__.py:
```diff
@@ -0,0 +1,35 @@
+        "Pedro M. Baeza <pedro.baeza@serviciosbaeza.com>",
+        "Ana Juaristi <ajuaristio@gmail.com>",
+        "Zakaria Makrelouf(Acsone) <z.makrelouf@gmail.com>",
+    ],
```
Contributors must go in the readme.

sbidoul committed:
align with v8

sbidoul committed:
[REF] refactoring of the onchanges

This should be easier to understand and cover all cases of
changing product template and product in any order, so you
always get a consistent list of attributes.

sbidoul committed:
[FIX] fix tests in purchase_product_variant_configurator

sbidoul commented:
I made some changes in https://github.com/OCA/product-variant/pull/44/commits/458e27aee3511686c4b0e30272ee2e0b6cdf67fb so the code is easier to compare with #38.

More important, in https://github.com/OCA/product-variant/pull/44/commits/8594867917526bc5bd7798f945ddc37f31435416 I did a big refactoring of the onchanges so the attributes manipulation is located in reusable and easy(er) to understand methods. And I restructered the product_tmpl and product onchanges so the attribute list always remain consistent, whatever the order in which the user changes the product/product template.

Now it feels much more robust and less surprising. Try it on the purchase order.

sbidoul committed:
Merge branch '9.0-variant-creation-warning-sbi' into 9.0-mig_purchase_product_variant_configurator

Commit:
Add some tests to the product_variant_configurator (#4)

sbidoul committed:
disable test of onchange_no_create_variant

because I don't know how to emulate a true onchange that
populates _origin

sbidoul committed:
pep8

sbidoul commented on product_variant_configurator/models/product_configurator.py:
```diff
@@ -0,0 +1,272 @@
+                _("You have to fill all the attributes values."))
+
+    @api.multi
+    def _create_variant_from_vals(self, vals):
```
@pedrobaeza this method _create_variant_from_vals is used nowhere. Do you remember it's purpose?

Garamotte replied:
@sbidoul This method is used in the stock_product_variants module, in odoomrp-wip.

sbidoul replied:
@pedrobaeza @sylvain-garancher would you agree to removed unused code from here and reintroduce it when needed?

Garamotte replied:
I have no idea about this : I just ran a `git grep _create_variant_from_vals` in the odoomrp-wip repository, and don't really know these modules :)

pedrobaeza replied:
It's used the same for stock_product_variants, as we can't create variants the regular way. Maybe we can use convert_to_write and convert_to_read methods for using a standard way.

sbidoul commented on purchase_product_variant_configurator/models/purchase_order_line.py:
```diff
@@ -0,0 +1,56 @@
+                lambda r: r.company_id.id == company_id)
+        )
+        self._suggest_quantity()
+        self._onchange_quantity()
```
@zakiuu these methods _suggest_quantity and _onchange_quantity do nothing if self.product_id is not set. That's why the quantity and price is not proposed.

Instead of replicating all this product-related logic for the product template, I think it's better to create the product earlier (as soon as all the attribute values are set - we can create a product in an onchange, right?)  and let the normal logic do it's work on the newly created product.

The only drawback (compared to creating the product upon purchase order confirmation) is that we may end up creating some variants that are not used because the user has changed some attribute values, but it is not a real problem in my opinion. And the logic will be so much simpler.

@pedrobaeza what do you think?

cc/ @chrisdec

pedrobaeza replied:
I prefer to not mess the concept of not creating product variants until the confirmation of the order. It's simple to implement the missing logic of that methods for the template, so better to go to that part.

sbidoul replied:
@pedrobaeza we the immediate creation in https://github.com/OCA/product-variant/pull/44/commits/645a1e5de9ca70b1a69638b5d6de5ba611f94ba4. It's easy and works very well. It is also much less surprising for the user. Finally preserving the existing logic based on product.product is not only much simpler, but allows to interoperate with other modules that would override it. 

I see many advantages, and only a small drawback. Can you please reconsider?

pedrobaeza replied:
Your solution doesn't work when you use _product_variant_types_ (to be renamed) and you have optional attributes, and you will get a lot of intermediate crap products. It also happens when you wrongly select the last attribute and change to another, that will continue creating a lot of non valid product.product. Even more, if you make a lot of quotations with possible variants, but they never reach the confirmed status, you will have to deal with all these created variants.

I prefer to stay as is and deal instead with the proper changes on methods. Other option is to create the product on memory, call upstream onchanges, and dispose it later, but this is very heavy for the performance.

zakmak88 committed:
[IMP] Create product when all the values are filled

sbidoul commented on purchase_product_variant_configurator/views/inherited_purchase_order_views.xml:
```diff
@@ -0,0 +1,55 @@
+                            type="object"
+                            icon="STOCK_COPY"
+                            attrs="{'invisible': [('state', 'not in', ['draft', 'sent'])]}"
+                            help="Duplicate line"/>
```
@zakiuu can you remove this duplicate button. IMO it's unrelated to this module.

zakmak88 committed:
[REF] Remove duplicate button

zakmak88 committed:
[REF] Remove print debug statements

zakmak88 committed:
[REF] PEP8

ghost commented:
the method _get_po_line_values_from_proc in procurement_order is not called anywhere while processing the purchase_order. it's copied I think from 8.0 and it's not useful in this case. if we want to add product template to purchase order we should do it elsewhere.

sbidoul committed:
[IMP] attribute prices are for sale only

plus, fix the visibility rule of the button,
to show the button if there are attributes on the product
(otherwise it would stay hidden until the first 2 variants
have been created)

sbidoul committed:
[FIX] typo in product attributes validity constraint

sbidoul committed:
[FIX] more robust configurator onchanges

sbidoul committed:
[FIX] return values for onchange_product_attribute_ids

zakiuu commented on product_variant_configurator/models/product_configurator.py:
```diff
@@ -36,28 +36,58 @@ def _compute_price_extra(self):
+            self.product_attribute_ids = attribute_lines
+
+    def _empty_attributes(self):
+        self.product_attribute_ids = self.product_attribute_ids.browse([])
```
@sbidoul we should delete those unneeded records.

zakmak88 committed:
[REF] Revert onchange_product_attribute_ids and make it behave exactly like onchange_product_id

zakmak88 committed:
[REF] remove unneeded import

zakmak88 committed:
[REF] a better display for attributes in the order line form

sbidoul commented on purchase_product_variant_configurator/views/inherited_purchase_order_views.xml:
```diff
@@ -9,38 +9,29 @@
-                <xpath expr="//field[@name='order_line']/tree//field[@name='product_id']"
-                       position="before">
-                    <field name="product_tmpl_id" />
-                </xpath>
```
@zakiuu product_tmpl_id must stay visible in the tree

sbidoul commented on purchase_product_variant_configurator/views/inherited_purchase_order_line_views.xml:
```diff
@@ -14,20 +14,6 @@
-            <field name="arch" type="xml">
-                <field name="product_id" position="before">
-                    <field name="product_tmpl_id" />
-                </field>
```
@zakiuu product_tmpl_id must stay visible in the tree

Garamotte commented on product_variant_configurator/models/product_product.py:
```diff
@@ -0,0 +1,105 @@
+    _name = "product.product"
+
+    # This is needed as the AbstractModel removes the delegated related field
+    name = fields.Char(related="product_tmpl_id.name")
```
@pedrobaeza Do you remember why the `name` field is redefined in the `product.configurator` model ?

If this is really needed, can it be possible to redefine this field using `compute` instead of `related` ?
The `related` attribute has precedence over `compute`, and it's not possible to empty it in v9 (this is now possible in v10).

Reason : I have a module that redefines the `name` field with `compute`, and don't find any other way to make both modules work together : I'd prefer to keep my custom `compute` behaviour, but even if my module depends on this one, the `related` overrides my `compute` method.

sbidoul commented:
The configurator on the purchase order line works fine.

It is broken on the product product form though. If someone wants to have a look at it, help is welcome.

zakmak88 committed:
[REF]Revert display of product template (removed by mistake)

mmalorni commented:
@sbidoul @zakiuu We have made 2 PR against your branch /acsone/product-variant/tree/9.0-mig_purchase_product_variant_configurator. 
It adds some tests and removes a function that is not longer necessary in v9.

Commit:
Add a test for onchange_product_attribute_ids and delete prodcurement_order (#5)

* Add a test for onchange_product_attribute_ids and delete prodcurement_order

* remove procurement_order from teh init file

sbidoul committed:
[DEL] remove old api method that is not used, untested and unmaintained

zakmak88 committed:
[FIX] Display product.product from the same way in the purchase order

zakmak88 committed:
[IMP] Add variants menu to the purchase module

zakmak88 committed:
[FIX] Test configuration validity before confirming the purchase order

sbidoul commented on product_variant_configurator/models/product_configurator.py:
```diff
@@ -185,8 +185,15 @@ def unlink(self):
+
     @api.model
-    def check_configuration_validity(self, vals):
+    def check_configuration_validity_from_vals(self, vals):
```
@zakiuu do we need these two variants of `check_configuration_validity`? It's annoying to have duplicated code.

zakiuu replied:
the check_configuration_validity_from_vals existed in the original code I didn't want to touch it plus it is not called anywhere, in the same time I wanted to  have a function to check the validity not from vals but for existing object (like order lines)

sbidoul replied:
Let's remove dead code, it can be reintroduced later if needed! Actually check_configurator_validity should not be needed as there is already the same check in https://github.com/OCA/product-variant/blob/cc88b0ce71bf907c4f300ef87f0cc03e59083c2f/product_variant_configurator/models/product_product.py#L81  so please revert this commit.

pedrobaeza replied:
You have also the methods `_convert_to_write` and `_convert_to_read` that can switch from dictionary to record and vice versa to handle things in a more homogeneous way and avoid duplicated code.

sbidoul commented on purchase_product_variant_configurator/views/inherited_product_product_views.xml:
```diff
@@ -0,0 +1,17 @@
+<openerp>
+    <data>
+        <record id="variant_purchase_action" model="ir.actions.act_window">
+            <field name="name">Variantes</field>
```
s/Variantes/Product Variants/

mmalorni commented:
@sbidoul I see you have branched to another PR #45. Has work stopped on this version or is it still valid? Are the last changes requested still valid?

sbidoul commented:
@mmalorni we have indeed switched to that branch. We very much prefer the explicit creation of the product.product when creating the `purchase.order.line` both from a technical standpoint (the code is much simpler), compatibility with Odoo core (where everything is designed to work with product.product) and usability we showed both to functional people and the prefer #45.

I also added variants on invoice lines there and its very simple.

I think all comments here have been attended.

So I'd think we can close this PR and continue on #45 (which is stabilizing at last...). 

@pedrobaeza what's your opinion? If you have use cases where the variant creation must absolutely be delayed until SO/PO confirmation it can still be done with the exact same abstract class, while hiding the "create product now" checkbox and adding the logic to create products on PO/SO confirmation in another module.

pedrobaeza commented:
The only doubt I have with this approach is if you will have all the features (pricelist, minimum quantity, variant extra price, etc), if you don't create the product.product record. If not, then it's not valid for me.

sbidoul commented:
@pedrobaeza the logic to make purchase.order.line work with an empty product_id can be added in additional modules. All the logic in product_variant_configurator remains exactly the same. So yes it's possible and clean.

pedrobaeza commented:
No, but I had it in v8 and I don't want to lose it or implement it in an additional module. It must be in this one.

sbidoul commented:
Doing both in the same module won't work.

What's wrong with an additional module named eg `purchase_create_variant_on_confirm` or something?

pedrobaeza commented:
I don't understand why they are not going to work. They were working on version 8. OK about including it in an additional module, but please don't strip out this functionality.

sbidoul commented:
Good. I'll try to shave some time to move that from this PR to #45. This beast is consuming a huge amount of time to get right. 

Are you ok with the modules names?
- product_variant_configurator
- purchase_product_variant_configurator
- purchase_create_variant_on_confirm
- invoice_product_variant_configurator

pedrobaeza commented:
The last 2 seem weird:
- `purchase_create_variant_on_confirm` should be connected with its parent module. What about `purchase_product_variant_configurator_on_confirm`. It's a bit large but very tight to the parent module.
- `invoice_product_variant_configurator` why do you need this module?

pedrobaeza commented:
Other option is to strip `product` word for all of them and let `variant_configurator` as the main conductor. It makes sense indeed:
- product_variant_configurator.
- purchase_variant_configurator.
- purchase_variant_configurator_on_confirm.
- sale_variant_configurator.
- ...

sbidoul commented:
> purchase_create_variant_on_confirm should be connected with its parent module. What about purchase_product_variant_configurator_on_confirm. It's a bit large but very tight to the parent module.

I'm fine with the long name if it's not too long for Odoo.

> invoice_product_variant_configurator why do you need this module?

Customer creating invoices without PO but willing to have the right product on the invoice nevertheless.

pedrobaeza commented:
What about my last proposition?

Last name should be then `account_variant_configurator` or `account_invoice_variant_configurator`.

sbidoul commented:
Okay, purchase_variant_configurator_on_confirm is on #45
